### PR TITLE
Create optimization_ideas.Rmd

### PR DESCRIPTION
**Pull request by saxenism**
Adding a corpus of possible optimization strategies for future rco optimizers. This vignette can be preview at: https://rpubs.com/hinduBale/strategies_ideas_rco

### CODE REVIEW

File: vignettes/potential-optimizers.Rmd [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+  %\VignetteEncoding{UTF-8}
+---
+
+# Potential optimizers
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+```
+
+## 1. Jump Threading
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: I would remove the numbering 1., 2., etc.

File: vignettes/potential-optimizers.Rmd [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+
+### Idea
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd:21]: New line after headings.

File: vignettes/potential-optimizers.Rmd [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+### Benchmark
+
+```{r jump_thread_benchmark, warning=FALSE, message=FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 100000
+autoplot(microbenchmark(jump(n), jump_opt(n)))
+```
+
+### Drawbacks
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: Don't put drawbacks if we can't find them.

File: vignettes/potential-optimizers.Rmd [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+
+  **To-Do**
+  
+## 2. Inline Expansion
+
+### Idea
+Replacing a function call site with the body of the called function is called Inline Expansion. This eliminates the function calling overhead and also the overhead of return call from a function. It also saves the overhead of variables push/pop on the stack while, function calling
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: ... is called "Inline Expansion".

Comment (jcrodriguez1989) [reply vignettes/potential-optimizers.Rmd]: ... stack while function calling.

File: vignettes/potential-optimizers.Rmd [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+### Drawbacks
+
+  **To-Do**
+  
+## 3. Memory Pre-Allocation
+
+### Idea
+
+### Code Examples
+As a general rule of thumb, in any programming language we should undertake the memory management as much as possible. When we grow a vector inside a loop, the vector asks the processor for extra space in between the running program and then proceeds, once it gets the required memory. This process is repeated for every iteration of the loop. Thus we should pre-allocate the required memory to a vector to avoid such delays.
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: this goes in the Idea section.

File: vignettes/potential-optimizers.Rmd [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+
+```{r pre_allocation_code_op}
+pre_mem_opt <- function(n) {
+  vec <- vector(mode = "numeric", length = n)
+  for (i in seq_len(n)) {
+    vec[i] <- i
+  }
+}
+
+pre_mem_opt_unknown_type <- function(n) {
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: Good idea, I would merge both as:
```r
pre_mem_opt <- function(n) {
  vec <- vector(length = n)
  for (i in seq_len(n)) {
    vec[i] <- i
  }
}
```

File: vignettes/potential-optimizers.Rmd [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 4. Vectorization
+
+### Idea
+A golden rule in R programming is to access the underlying C/Fortran routines as quickly as possible; the fewer functions calls required to achieve this, the better.Many R functions are therefore vectorised, that is the function’s inputs and/or outputs naturally work with vectors, reducing the number of function calls required.
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: ...better. Many...

File: vignettes/potential-optimizers.Rmd [hunk 8]
```diff
@@ -0,0 +1,670 @@
...
+library("microbenchmark")
+n <- 10000
+autoplot(microbenchmark(non_vectorized(n), vectorized(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 5. Caching
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: It requires using a non-base library. We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+library("microbenchmark")
+n <- 20
+autoplot(microbenchmark(fibR(n), mem_fibR(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 6. Reversing elements
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: It is too specific. It is optimizing an algorithm, not an optimizer technique. We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+library("microbenchmark")
+n <- 1000
+autoplot(microbenchmark(reverse_fun(n), reverse_fun_opt(n), times = 10))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 7. Efficient Sorting
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: Same comment as the previous topic. We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 11]
```diff
@@ -0,0 +1,670 @@
...
+library("microbenchmark")
+n <- 100000
+autoplot(microbenchmark(sort_fun(n), sort_fun_opt(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 8. min/max using which function
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: Same as before, `rco` should focus on optimizing code chunks, not algorithms.
I mean, `rco` should reorganize code, resulting in more efficient code.
We should remove this idea for now.

File: vignettes/potential-optimizers.Rmd [hunk 12]
```diff
@@ -0,0 +1,670 @@
...
+library("microbenchmark")
+n <- 1000
+autoplot(microbenchmark(which_fun(n), which_fun_opt(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 9. Row and Column Operations
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 13]
```diff
@@ -0,0 +1,670 @@
...
+library("ggplot2")
+library("microbenchmark")
+autoplot(microbenchmark(rowCol_fun(), rowCol_fun_opt()))
+```
+
+### Drawbacks
+ 
+  **To-Do**
+  
+## 10. Detect NA values
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 14]
```diff
@@ -0,0 +1,670 @@
...
+library("microbenchmark")
+n <- 10000
+autoplot(microbenchmark(detectNA(n), detectNA_opt(n)))
+```
+
+### Drawbacks
+  
+  **To-Do**
+  
+## 11. Matrix v/s Data Frames(Lists)
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 15]
```diff
@@ -0,0 +1,670 @@
...
+library("ggplot2")
+library("microbenchmark")
+autoplot(microbenchmark(df_fun(), mat_fun()))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 12. Use of Native Binary Formats for I/O
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: We should remove it for now.

File: vignettes/potential-optimizers.Rmd [hunk 16]
```diff
@@ -0,0 +1,670 @@
...
+  mtcars$carb,
+  mtcars[[c(11)]],
+  mtcars[[11]],
+  .subset2(mtcars, 11)
+))
+```
+
+### Drawbacks
+
+  **To-Do**
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: We had the drawbacks for this, you should add it.

File: vignettes/potential-optimizers.Rmd [hunk 17]
```diff
@@ -0,0 +1,670 @@
...
+  mtcars[[c(11, 32)]],
+  mtcars[[11]][32],
+  .subset2(mtcars, 11)[32],
+  times = 1000L
+))
+```
+
+### Drawbacks
+
+  **To-Do**
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: Also here, we have it.

File: vignettes/potential-optimizers.Rmd [hunk 18]
```diff
@@ -0,0 +1,670 @@
...
+
+## External packages that can be explored for further speed-ups
```
Comment (jcrodriguez1989) [inline vignettes/potential-optimizers.Rmd]: I would keep following lines just as comments for us, remove them from the vignette.
We could get ideas from these packages.

**Review by jcrodriguez1989 — CHANGES_REQUESTED** [review]
We should remove ideas from 5 to 12 as they are not optimizing code chunks, they are using other functions (and we can't make sure what a function does, even if we know the name). If there is a `mean` function in the code, it is possible that previously there is something like `mean <- function() NULL`, so don't trust function names.

As it is organized the project now, using Rstudio, you can open the project, install the `pkgdown` package, and type `pkgdown::build_site()` to see how the vignette/site is being rendered.

File: vignettes/potential-optimizers.Rmd [hunk 19]
```diff
@@ -133,29 +133,26 @@ pre_mem <- function(n) {
...
 n <- 100000
-autoplot(microbenchmark(pre_mem(n), pre_mem_opt(n), pre_mem_opt_unknown_type(n)))
+autoplot(microbenchmark(mem_alloc(n), mem_alloc_opt(n)))
 ```
 
 ## Vectorization
 
 ### Idea
 
-A golden rule in R programming is to access the underlying C/Fortran routines as quickly as possible; the fewer functions calls required to achieve this, the better. Many R functions are therefore vectorised, that is the function’s inputs and/or outputs naturally work with vectors, reducing the number of function calls required.
```
Comment (saxenism) [inline vignettes/potential-optimizers.Rmd]: Hello Sir, I wrote this based on my readings [from here](https://csgillespie.github.io/efficientR/programming.html#general). I realize that what I have written may seem wrong without context, so have a look at the above URL. Go to section 3.2 *General Advice* and in that you'll find the *golden rule* as mentioned by the author. Thanks.

File: vignettes/potential-optimizers.Rmd [hunk 20]
```diff
@@ -238,44 +206,19 @@ autoplot(microbenchmark(
 
 ### Drawbacks
 
-
 1. For some R classes, the `[[ ]]` operator and `.subset` work differently. For instance, they seem to be equivalent for `data.frame` but are not the same for `matrix` class.
 
-2. Moreover, both [[ ]] and .subset2  are functions and in R, any function can be overwritten. Thus the above optimization can be made to fail just by redefining, say, the the `.subset2` function.
+2. Moreover, both `[[ ]]` and `.subset2`  are functions and in R, any function can be overwritten. Thus the above optimization can be made to fail just by redefining, say, the the `.subset2` function.
```
Comment (saxenism) [inline vignettes/potential-optimizers.Rmd:211]: Hello Sir, it's written,  *the the `.subset2` function.* , but, I'll take care of that. No issues.

**Review by saxenism — COMMENTED** [review]
I just made 2 comments. Please have a look at that. Apart from that it seems pretty good to me. Cheers :D

**Review by saxenism — COMMENTED** [review]
All good 👍

File: vignettes/potential-optimizers.Rmd (hunk 20 above)
Comment (jcrodriguez1989) [reply vignettes/potential-optimizers.Rmd:211]: Please :)

**Pull request merged**